File: ios/oplayer/Controller/Gateway/RuDEX.m
//
//  RuDEX.m
//  oplayer
//
//  Created by SYALON on 13-11-20.
//
//

#import "RuDEX.h"
#import "OrgUtils.h"

@interface RuDEX()
{
}

@end

@implementation RuDEX

- (void)dealloc
{
}

- (WsPromise*)queryCoinList
{
    id api_base = [self.api_config_json objectForKey:@"base"];
    id coin_list = [self.api_config_json objectForKey:@"coin_list"];
    assert(coin_list);
    
    return [OrgUtils asyncFetchUrl:[NSString stringWithFormat:@"%@%@", api_base, coin_list] args:nil];
}

- (NSArray*)processCoinListData:(NSArray*)data_array balanceHash:(NSDictionary*)balanceHash
{
    //{
    //    backingCoin = PPY;
    //    confirmations =     {
    //        type = irreversible;
    //    };
    //    depositAllowed = 1;
    //    description = "PeerPlays currency";
    //    gatewayWallet = "rudex-gateway";
    //    issuer = "rudex-ppy";
    //    issuerId = "1.2.353611";
    //    memoSupport = 1;
    //    minAmount = 20000;
    //    name = Peerplays;
    //    precision = 5;
    //    symbol = PPY;
    //    walletType = peerplays;
    //    withdrawalAllowed = 1;
    //},
    NSMutableArray* result = [NSMutableArray array];
    if (data_array && [data_array isKindOfClass:[NSArray class]] && [data_array count] > 0){
        for (id item in data_array) {
            BOOL enableDeposit = [[item objectForKey:@"depositAllowed"] boolValue];
            BOOL enableWithdraw = [[item objectForKey:@"withdrawalAllowed"] boolValue];
            NSDecimalNumber* n_minAmount = [NSDecimalNumber decimalNumberWithMantissa:[item[@"minAmount"] unsignedLongLongValue]
                                                                             exponent:-[item[@"precision"] integerValue]
                                                                           isNegative:NO];
            //  网络确认数
            NSString* confirm_block_number = @"";
            id confirmations = [item objectForKey:@"confirmations"];
            if (confirmations && [[[confirmations objectForKey:@"type"] lowercaseString] isEqualToString:@"blocks"]){
                confirm_block_number = [NSString stringWithFormat:@"%@", [confirmations objectForKey:@"value"]];
            }
            
            NSString* symbol = [item[@"symbol"] uppercaseString];
            assert(symbol);
            assert(balanceHash);
            id balance_item = [balanceHash objectForKey:symbol] ?: @{@"iszero":@YES};
            
            GatewayAssetItemData* appext = [[GatewayAssetItemData alloc] init];
            appext.enableWithdraw = enableWithdraw;
            appext.enableDeposit = enableDeposit;
            appext.symbol = symbol;
            appext.backSymbol = [item[@"backingCoin"] uppercaseString];
            appext.name = item[@"name"];
            appext.intermediateAccount = item[@"issuerId"] ?: item[@"issuer"];
            appext.balance = balance_item;
            appext.depositMinAmount = [NSString stringWithFormat:@"%@", n_minAmount];
            appext.withdrawMinAmount = [NSString stringWithFormat:@"%@", n_minAmount];
            appext.withdrawGateFee = @"";
            appext.supportMemo = [[item objectForKey:@"memoSupport"] boolValue];
            appext.confirm_block_number = confirm_block_number;
            appext.coinType = item[@"symbol"];
            appext.backingCoinType = item[@"backingCoin"];
            
            id new_item = [item mutableCopy];
            [new_item setObject:appext forKey:@"kAppExt"];
            [result addObject:[new_item copy]];
        }
    }
    return result;
}

- (WsPromise*)requestDepositAddress:(id)item fullAccountData:(id)fullAccountData vc:(VCBase*)vc
{
    assert(item);
    GatewayAssetItemData* appext = [item objectForKey:@"kAppExt"];
    assert(appext);
    
    id account_data = [fullAccountData objectForKey:@"account"];
    
    //  if memo not supported - should request deposit address
    if (![[item objectForKey:@"memoSupport"] boolValue]){
        id walletType = [item objectForKey:@"walletType"];
        assert(walletType);
        id request_deposit_address_base = [self.api_config_json objectForKey:@"request_deposit_address"];
        id final_url = [NSString stringWithFormat:@"%@%@", self.api_config_json[@"base"],
                        [NSString stringWithFormat:request_deposit_address_base, walletType]];
        
        return [self requestDepositAddressCore:item
                                        appext:appext
                   request_deposit_address_url:final_url
                             full_account_data:fullAccountData
                                            vc:vc];
    }
    
    //  if support memo (memo is fixed now = dex:bitshares-account-name eg. dex:btsacc)
    NSString* inputAddress = nil;
    id gatewayWallet = [item objectForKey:@"gatewayWallet"];
    if (gatewayWallet && ![gatewayWallet isEqualToString:@""]){
        inputAddress = gatewayWallet;
    }
    id account_name = account_data[@"name"];
    NSString* inputMemo = [NSString stringWithFormat:@"dex:%@", account_name];
    return [WsPromise promise:^(WsResolveHandler resolve, WsRejectHandler reject) {
        if (inputAddress){
            id depositItem = @{
                               @"inputAddress":inputAddress,
                               @"inputCoinType":[appext.backingCoinType lowercaseString],
                               @"inputMemo":inputMemo,
                               @"outputAddress":account_name,
                               @"outputCoinType":[appext.coinType lowercaseString],
                               };
            resolve(depositItem);
        }else{
            resolve(NSLocalizedString(@"kVcDWErrTipsRequestDepositAddrFailed", @"获取充币地址异常，请联系网关客服。"));
        }
    }];
}

/**
 *  验证地址、备注、数量是否有效
 */
- (WsPromise*)checkAddress:(id)item address:(NSString*)address memo:(NSString*)memo amount:(NSString*)amount
{
    //  TODO:仅验证地址
    
    id walletType = [item objectForKey:@"walletType"];
    assert(walletType);
    
    id check_address_base = [self.api_config_json objectForKey:@"check_address"];
    id final_url = [NSString stringWithFormat:@"%@%@", self.api_config_json[@"base"],
                    [NSString stringWithFormat:check_address_base, walletType]];
    
    return [WsPromise promise:^(WsResolveHandler resolve, WsRejectHandler reject) {
        id post_args = @{
                         @"address":address,
                         };
        [[[OrgUtils asyncPostUrl_jsonBody:final_url args:post_args] then:(^id(id resp_data) {
            if (resp_data && [resp_data isKindOfClass:[NSDictionary class]] && [[resp_data objectForKey:@"isValid"] boolValue]){
                resolve(@YES);
            }else{
                resolve(@NO);
            }
            return nil;
        })] catch:(^id(id error) {
            resolve(@NO);
            return nil;
        })];
    }];
}

@end

File: android/app/src/main/java/com/btsplusplus/fowallet/gateway/RuDEX.kt
package com.btsplusplus.fowallet.gateway

import android.content.Context
import bitshares.*
import com.btsplusplus.fowallet.R
import org.json.JSONArray
import org.json.JSONObject

class RuDEX : GatewayBase() {

    override fun queryCoinList(): Promise {
        val api_base = _api_config_json.getString("base")
        val coin_list = _api_config_json.getString("coin_list")

        return OrgUtils.asyncJsonGet("$api_base$coin_list")
    }

    override fun processCoinListData(data_array: JSONArray, balanceHash: JSONObject): JSONArray? {
        //{
        //    backingCoin = PPY;
        //    confirmations =     {
        //        type = irreversible;
        //    };
        //    depositAllowed = 1;
        //    description = "PeerPlays currency";
        //    gatewayWallet = "rudex-gateway";
        //    issuer = "rudex-ppy";
        //    issuerId = "1.2.353611";
        //    memoSupport = 1;
        //    minAmount = 20000;
        //    name = Peerplays;
        //    precision = 5;
        //    symbol = PPY;
        //    walletType = peerplays;
        //    withdrawalAllowed = 1;
        //},
        val result = JSONArray()
        for (it in data_array.forin<JSONObject>()) {
            val item = it!!
            val enableDeposit = item.isTrue("depositAllowed")
            val enableWithdraw = item.isTrue("withdrawalAllowed")
            val n_minAmount = bigDecimalfromAmount(item.getString("minAmount"), item.getInt("precision"))

            //  网络确认数
            var confirm_block_number = ""
            val confirmations = item.optJSONObject("confirmations")
            if (confirmations != null && confirmations.getString("type").equals("blocks", true)) {
                confirm_block_number = confirmations.getString("value")
            }

            val symbol = item.getString("symbol").toUpperCase()
            val balance_item = balanceHash.optJSONObject(symbol)
                    ?: jsonObjectfromKVS("iszero", true)

            val appext = GatewayAssetItemData()
            appext.enableWithdraw = enableWithdraw
            appext.enableDeposit = enableDeposit
            appext.symbol = symbol
            appext.backSymbol = item.getString("backingCoin").toUpperCase()
            appext.name = item.getString("name")
            appext.intermediateAccount = item.optString("issuerId") ?: item.optString("issuer")
            appext.balance = balance_item
            appext.depositMinAmount = n_minAmount.toString()
            appext.withdrawMinAmount = n_minAmount.toString()
            appext.withdrawGateFee = ""
            appext.supportMemo = item.getBoolean("memoSupport")
            appext.confirm_block_number = confirm_block_number
            appext.coinType = item.getString("symbol")
            appext.backingCoinType = item.getString("backingCoin")

            item.put("kAppExt", appext)
            result.put(item)
        }
        return result
    }

    /**
     *  请求充值地址
     */
    override fun requestDepositAddress(item: JSONObject, fullAccountData: JSONObject, ctx: Context): Promise {
        val appext = item.get("kAppExt") as GatewayAssetItemData
        val account_data = fullAccountData.getJSONObject("account")

        //  if memo not supported - should request deposit address
        if (!item.isTrue("memoSupport")) {
            val walletType = item.getString("walletType")
            val request_deposit_address_base = _api_config_json.getString("request_deposit_address")
            val final_url = "${_api_config_json.getString("base")}${String.format(request_deposit_address_base, walletType)}"
            return requestDepositAddressCore(item, appext, final_url, fullAccountData, ctx)
        }

        //  if support memo (memo is fixed now = dex:bitshares-account-name eg. dex:btsacc)
        var inputAddress: String? = null
        val gatewayWallet = item.optString("gatewayWallet")
        if (gatewayWallet != "") {
            inputAddress = gatewayWallet
        }
        val account_name = account_data.getString("name")
        val inputMemo = "dex:$account_name"
        val p = Promise()
        if (inputAddress != null) {
            val depositItem = JSONObject().apply {
                put("inputAddress", inputAddress)
                put("inputCoinType", appext.backingCoinType.toLowerCase())
                put("inputMemo", inputMemo)
                put("outputAddress", account_name)
                put("outputCoinType", appext.coinType.toLowerCase())
            }
            p.resolve(depositItem)
        } else {
            p.resolve(R.string.kVcDWErrTipsRequestDepositAddrFailed.xmlstring(ctx))
        }
        return p
    }

    /**
     *  验证地址是否有效
     */
    override fun checkAddress(item: JSONObject, address: String, memo: String?, amount: String): Promise {

        //  TODO:仅验证地址
        val walletType = item.getString("walletType")

        val check_address_base = _api_config_json.getString("check_address")
        val api_base = _api_config_json.getString("base")
        val final_url = "$api_base${String.format(check_address_base, walletType)}"

        val p = Promise()
        OrgUtils.asyncPost_jsonBody(final_url, jsonObjectfromKVS("address", address)).then {
            val json = it as? JSONObject
            if (json != null && json.isTrue("isValid")) {
                p.resolve(true)
            } else {
                p.resolve(false)
            }
            return@then null
        }.catch {
            p.resolve(false)
        }
        return p
    }
}